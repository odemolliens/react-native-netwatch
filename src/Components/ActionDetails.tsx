import * as React from 'react';
import { useContext } from 'react';
import { View, StyleSheet } from 'react-native';
import { Subheading } from 'react-native-paper';
import { reduxTag } from './Status';
import { getTime, getShortDate, isLongText, addEllipsis } from '../Utils/helpers';
import { Text, Title } from './Text';
import { ThemeContext } from '../Theme';
import ReduxAction from '../Core/Objects/ReduxAction';
import { ViewMoreButton } from './ViewMoreButton';

export interface IProps {
  testId?: string;
  item: ReduxAction;
  reduxAction: any;
  onPressViewMore: (value: boolean) => void;
}

export const ActionDetails: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <View style={{ flex: 1, width: '100%' }}>
      {/* REDUX + REDUX TAG */}
      <View style={[styles.line]}>
        <Title style={[{ marginRight: 6 }, { color: theme.textColorOne }]}>REDUX</Title>
        {reduxTag()}
      </View>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Started at : </Text>
        <Text>{`${getShortDate(props.item.startTime)} - ${getTime(props.item.startTime)}`}</Text>
      </View>

      <Subheading
        style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
      >
        Content
      </Subheading>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Type</Text>
      </View>
      <View style={[styles.line]}>
        <Text>{props.item.action.type}</Text>
      </View>

      <View style={[styles.line]}>
        {props.reduxAction.label && (
          <Text style={{ color: theme.textColorFour }}>{`${props.reduxAction.label
            .charAt(0)
            .toUpperCase()}${props.reduxAction.label.slice(1)} :`}</Text>
        )}
      </View>
      <View style={[styles.line]}>
        <View style={[styles.attribtuesContainer, { flex: 1, paddingHorizontal: 0 }]}>
          {props.reduxAction.payload && <Text>{addEllipsis(JSON.stringify(props.reduxAction.payload, null, 2))}</Text>}
          {isLongText(addEllipsis(JSON.stringify(props.reduxAction.payload, null, 2))) && (
            <View style={{ alignItems: 'flex-end' }}>
              <ViewMoreButton onPress={() => props.onPressViewMore(true)} />
            </View>
          )}
        </View>
      </View>
    </View>
  );
};

export default ActionDetails;

const styles = StyleSheet.create({
  line: {
    flexDirection: 'row',
    paddingHorizontal: 16,
    marginBottom: 10,
    alignItems: 'center',
  },

  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },

  attribtuesContainer: {
    justifyContent: 'flex-start',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
});
