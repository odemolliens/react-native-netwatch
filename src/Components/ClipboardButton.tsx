import * as React from 'react';
import { useContext } from 'react';
import { TouchableOpacity, Text, StyleSheet } from 'react-native';
import { ThemeContext } from '../Theme';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';

export interface IProps {
  onPress: () => void;
}

// Button use in the details screens to copy to clipboard a text. It's just a button text.
export const ClipboardButton: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);
  return (
    <TouchableOpacity testID="buttonCopyToClipboard" style={styles.container} onPress={props.onPress}>
      <MaterialCommunityIcons name="clipboard-arrow-left-outline" color={theme.primaryColor} size={14} />
      <Text style={[{ color: theme.primaryColor, marginLeft: 6 }]}>Copy</Text>
    </TouchableOpacity>
  );
};

export default ClipboardButton;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    borderLeftWidth: 0,
    alignItems: 'center',
  },
});
