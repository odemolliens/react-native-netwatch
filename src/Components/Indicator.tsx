import React, { useState, useRef, useEffect, useContext } from 'react';
import { Animated, TouchableOpacity, StyleSheet } from 'react-native';
import { ThemeContext } from '../Theme';
import { Title } from './Text';

export interface IProps {
  success: number;
  warning: number;
  failed: number;
}

// This component is a bar divided in three parts
// A part to represent success request, a part for warning and a part
// for requests failed.
// It is possible to press on the indicator to expand it and displayed
// a value in percent for each part.
export const Indicator: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);
  const height = useRef(new Animated.Value(6)).current;
  const [isOpened, setIsOpened] = useState<boolean>(false);

  useEffect(() => {
    Animated.timing(height, {
      toValue: isOpened ? 24 : 6,
      duration: 500,
      useNativeDriver: false,
    }).start();
  }, [isOpened]);

  return (
    <TouchableOpacity testID="openStatsButton" style={{ flexDirection: 'row' }} onPress={() => setIsOpened(!isOpened)}>
      <Animated.View
        key="success"
        style={{
          height,
          width: `${props.success.toString()}%`,
          backgroundColor: theme.successColor,
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {isOpened && props.success > 10 && (
          <Title style={[styles.text, { color: theme.textColorOne }]}>{`${Math.round(props.success)}%`}</Title>
        )}
      </Animated.View>
      <Animated.View
        key="warning"
        style={{
          height,
          width: `${props.warning.toString()}%`,
          backgroundColor: theme.warningColor,
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {isOpened && props.warning > 10 && (
          <Title style={[styles.text, { color: theme.textColorOne }]}>{`${Math.round(props.warning)}%`}</Title>
        )}
      </Animated.View>
      <Animated.View
        key="failure"
        style={{
          height,
          width: `${props.failed.toString()}%`,
          backgroundColor: theme.failedColor,
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {isOpened && props.failed > 10 && (
          <Title style={[styles.text, { color: theme.textColorOne }]}>{`${Math.round(props.failed)}%`}</Title>
        )}
      </Animated.View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  text: {
    fontSize: 16,
    fontWeight: 'bold',
  },
});

export default Indicator;
