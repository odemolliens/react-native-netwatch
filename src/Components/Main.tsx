import * as React from 'react';
import { useState, useEffect, useContext, useRef } from 'react';
import { View, StyleSheet, Alert, TouchableOpacity, FlatList, Share, Keyboard } from 'react-native';
import { Appbar, Searchbar } from 'react-native-paper';
import Item from './Item';
import FeatherIcon from 'react-native-vector-icons/Feather';
import { Settings } from './Settings';
import { ILog, SourceType, RequestMethod, EnumSourceType, EnumFilterType } from '../types';
import RNRequest from '../Core/Objects/RNRequest';
import ReduxAction from '../Core/Objects/ReduxAction';
import NRequest from '../Core/Objects/NRequest';
import { ThemeContext } from '../Theme';

export interface IProps {
  testId?: string;
  onPress: Function;
  onPressClose: (value: boolean) => void;
  onPressDetail: (value: boolean) => void;
  visible?: boolean;
  reduxActions: ReduxAction[];
  rnRequests: RNRequest[];
  nRequests: NRequest[];
  clearAll: Function;
  maxRequests?: number;
}

const _positionFlatList = 0;

export const Main = (props: IProps) => {
  const theme = useContext(ThemeContext);
  const refFlatList = useRef<FlatList<any>>(null);
  const [flatListLastOffset, setFlatListLastOffset] = useState<number>(0);
  const [requests, setRequests] = useState<Array<ILog>>([]);
  const [searchQuery, setSearchQuery] = useState<string>('');
  const [source, setSource] = useState<SourceType | EnumSourceType>(EnumSourceType.All);
  const [filter, setFilter] = useState<RequestMethod | EnumFilterType>(EnumFilterType.All);
  const [settingsVisible, setSettingsVisible] = useState(false);
  const [deleteVisible, setDeleteVisible] = useState<boolean>(false);
  const onChangeSearch = (query: string) => setSearchQuery(query);
  const hideDialogAndDelete = () => {
    clearList();
    if (deleteVisible) setDeleteVisible(false);
  };

  useEffect(() => {
    if (source === EnumSourceType.Redux) {
      if (filter !== EnumFilterType.All) setFilter(EnumFilterType.All);
      setRequests(props.reduxActions);
    } else if (source === EnumSourceType.ReactNativeRequest) {
      setRequests(props.rnRequests);
    } else if (source === EnumSourceType.Nativerequest) {
      setRequests(props.nRequests.sort(compare).reverse());
    } else {
      setRequests(mergeArrays(props.reduxActions, props.rnRequests, props.nRequests).sort(compare).reverse());
    }

    // TODO: Keep the same color after update for each items
    // TODO: We can add many items during one loop -> we must set the offset to handle that
    // TODO: What if we filtering -> scroll to 0 ?
    if (_positionFlatList > 55) {
      setFlatListLastOffset(_positionFlatList);
    }
  }, [props.rnRequests, props.reduxActions, props.nRequests, source, filter]);

  useEffect(() => {
    refFlatList?.current?.scrollToOffset({
      animated: false,
      offset: flatListLastOffset + 60 * 1,
    });
  }, [flatListLastOffset]);

  const mergeArrays = (...arrays: Array<ILog[]>) => {
    return [...arrays.flat()];
  };

  const compare = (a: ILog, b: ILog) => {
    const startTimeA = a.startTime;
    const startTimeB = b.startTime;

    let comparison = 0;
    if (startTimeA > startTimeB) {
      comparison = 1;
    } else if (startTimeA < startTimeB) {
      comparison = -1;
    }
    return comparison;
  };

  const filteredRequests = (): Array<any> =>
    requests.slice(0, props.maxRequests).filter(request => {
      if (filter === EnumFilterType.All) {
        if (request instanceof ReduxAction) {
          return JSON.stringify(request.action).toLowerCase().includes(searchQuery.toLowerCase());
        }
      }
      if (request instanceof RNRequest || request instanceof NRequest) {
        if (request.type === EnumSourceType.ReactNativeRequest) {
          if (filter === request.method || filter === EnumFilterType.All) {
            return (
              request.type === EnumSourceType.ReactNativeRequest &&
              (request.url?.toLowerCase().includes(searchQuery.toLowerCase()) ||
                request.status === parseInt(searchQuery, 10) ||
                request.method.toLowerCase() === searchQuery.toLowerCase())
            );
          }
        } else {
          if (filter === request.method || filter === EnumFilterType.All) {
            return (
              request.type === EnumSourceType.Nativerequest &&
              (request.url?.toLowerCase().includes(searchQuery.toLowerCase()) ||
                request.status === parseInt(searchQuery, 10) ||
                request.method.toLowerCase() === searchQuery.toLowerCase())
            );
          }
        }
      }
      return false;
    });

  const clearList = () => {
    props.clearAll();
    setRequests([]);
  };

  const formatSharedMessage = (array: Array<ILog>): string => {
    const _report = array.map(item => {
      return Object.values(item).flat().join(';');
    });
    return _report.join('\n');
  };

  const onShare = async (): Promise<void> => {
    try {
      await Share.share({
        message: formatSharedMessage(requests),
      });
    } catch (error) {
      Alert.alert('Error', error.message);
    }
  };

  const onDelete = () => {
    Alert.alert('Warning', 'Do you really want clear the call list?', [
      {
        text: 'Cancel',
        onPress: () => setDeleteVisible(false),
        style: 'cancel',
      },
      { text: 'OK', onPress: hideDialogAndDelete },
    ]);
  };

  const _renderItems = (_props: { item: ReduxAction | RNRequest | NRequest; index: number }) => {
    return (
      <Item
        testID={`${_props.index}`}
        item={_props.item}
        onPress={() => {
          props.onPress(_props.item);
          props.onPressDetail(true);
        }}
        color={theme.gray800}
      />
    );
  };

  return (
    <>
      <Appbar.Header style={[styles.header, { backgroundColor: theme.gray900 }]}>
        <TouchableOpacity style={[styles.button, { borderLeftWidth: 0 }]}>
          <FeatherIcon
            name="x"
            color={theme.white}
            size={24}
            onPress={() => (settingsVisible ? setSettingsVisible(false) : props.onPressClose(false))}
          />
        </TouchableOpacity>
        <Appbar.Content color={theme.blue500} title="Netwatch" titleStyle={{ fontSize: 18 }} />
        <TouchableOpacity style={[styles.button, { borderLeftWidth: 0 }]}>
          <FeatherIcon name="download" color={theme.white} size={24} onPress={onShare} />
        </TouchableOpacity>
      </Appbar.Header>
      <View style={[styles.options, { backgroundColor: theme.gray800 }]}>
        <Searchbar
          onFocus={() => {
            setSettingsVisible(false);
          }}
          onChange={() => {
            setSettingsVisible(false);
          }}
          placeholderTextColor={theme.gray400}
          iconColor={theme.gray500}
          inputStyle={{ color: theme.gray400 }}
          style={[styles.searchBar, { backgroundColor: theme.gray800 }]}
          placeholder={'Search'}
          onChangeText={onChangeSearch}
          value={searchQuery}
        />

        <TouchableOpacity
          testID="showDetailsButton"
          style={[
            styles.button,
            {
              borderLeftColor: theme.gray900,
              backgroundColor: settingsVisible ? theme.gray900 : theme.gray800,
            },
          ]}
          onPress={() => {
            Keyboard.dismiss();
            setSettingsVisible(!settingsVisible);
          }}
        >
          <FeatherIcon
            name="filter"
            color={filter !== EnumFilterType.All || source !== EnumSourceType.All ? theme.blue500 : theme.gray300}
            size={24}
          />
        </TouchableOpacity>

        <TouchableOpacity
          testID="deleteListButton"
          style={[styles.button, { borderLeftColor: theme.gray900 }]}
          onPress={onDelete}
        >
          <FeatherIcon name="trash-2" color={theme.gray300} size={24} />
        </TouchableOpacity>
      </View>

      {settingsVisible ? (
        <Settings source={source} onSetSource={setSource} filter={filter} onSetFilter={setFilter} />
      ) : (
        <FlatList
          ref={refFlatList}
          maintainVisibleContentPosition={{
            autoscrollToTopThreshold: 10,
            minIndexForVisible: 1,
          }}
          style={{ backgroundColor: theme.gray800 }}
          renderItem={_renderItems}
          keyExtractor={item => `${item._id.toString()}${item.startTime}${item.type}`}
          data={filteredRequests()}
        />
      )}
    </>
  );
};

const styles = StyleSheet.create({
  header: {
    elevation: 0,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0,
  },
  searchBar: {
    flex: 1,
    elevation: 0,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0,
  },
  options: {
    height: 56,
    flexDirection: 'row',
    alignItems: 'center',
  },
  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
});
