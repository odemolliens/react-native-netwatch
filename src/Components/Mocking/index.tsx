import * as React from 'react';
import { useEffect } from 'react';
import { EditMockResponse } from './EditMockResponse';
import { MockList } from './MockList';
import { View } from 'react-native';
import { MockResponse } from './utils';

export function MockingNavigator(props: {
  mockResponse?: MockResponse;
  onPressBack: (showDetails: boolean) => void;
  update: boolean;
}) {
  const [showList, setShowList] = React.useState(false);
  const [showEdit, setShowEdit] = React.useState(false);
  const [mockResponse, setMockResponse] = React.useState<MockResponse | undefined>();
  const [update, setUpdate] = React.useState(false);

  useEffect(() => {
    setShowList(!props.mockResponse);
    setShowEdit(!!props.mockResponse);
    setMockResponse(props.mockResponse);
    setUpdate(props.update);
  }, [props.mockResponse, props.update]);

  return (
    <View style={{ flex: 1 }}>
      {showList && (
        <MockList
          onPressBack={() => props.onPressBack(false)}
          showEdit={mr => {
            setShowEdit(true);
            setMockResponse(mr);
            setUpdate(true);
          }}
        />
      )}
      {showEdit && mockResponse && (
        <EditMockResponse
          update={update}
          mockResponse={mockResponse}
          onPressBack={() => {
            setShowList(true);
            setShowEdit(false);
          }}
        />
      )}
    </View>
  );
}
