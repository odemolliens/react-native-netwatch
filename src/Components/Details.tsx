import * as React from 'react';
import { useContext, useState } from 'react';
import { StyleSheet, View, ScrollView, Alert, TouchableOpacity } from 'react-native';
import Share from 'react-native-share';
import { Appbar, Snackbar } from 'react-native-paper';
import RNRequest from '../Core/Objects/RNRequest';
import NRequest from '../Core/Objects/NRequest';
import ReduxAction from '../Core/Objects/ReduxAction';
import FeatherIcon from 'react-native-vector-icons/Feather';
import { ThemeContext } from '../Theme';
import { ActionDetails } from './ActionDetails';
import JSONDetails from './JSONDetails';
import {
  getGeneralElementsAsArray,
  getRequestHeadersElementsAsArray,
  getResponseHeadersElementsAsArray,
  formatSharedMessage,
} from '../Utils/helpers';
import { RequestDetails } from './RequestDetails';
import { ILog } from '../types';

export interface IProps {
  testId?: string;
  onPressBack: (showDetails: boolean) => void;
  item: ILog;
}

export const Details: React.FC<IProps> = props => {
  const theme = useContext(ThemeContext);
  const [snackBarVisibility, setSnackBarVisibility] = useState<boolean>(false);
  const [snackBarMessage, setSnackBarMessage] = useState<string>('');
  const [showJSONResponseDetails, setShowJSONResponseDetails] = useState<boolean>(false);
  const [showJSONRequestDetails, setShowJSONRequestDetails] = useState<boolean>(false);
  const [showJSONActionDetails, setShowJSONActionDetails] = useState<boolean>(false);
  let _content = null;
  let _action: any = () => {};

  // _onShareRequest and _onShareReduxAction open the share panel. The text format is
  // plain text
  const _onShareRequest = async (item: RNRequest | NRequest): Promise<void> => {
    try {
      await Share.open({
        message: formatSharedMessage(
          getGeneralElementsAsArray(item),
          getRequestHeadersElementsAsArray(item),
          item.dataSent,
          getResponseHeadersElementsAsArray(item),
          item.response,
        ),
      });
    } catch (error) {
      Alert.alert('Error', error.message);
    }
  };

  const _onShareReduxAction = async (item: ReduxAction): Promise<void> => {
    const _type = item.action.type.toUpperCase();
    const _payload = JSON.stringify(item.action.payload, null, 2);
    try {
      await Share.open({
        message: `${_type}\n${_payload}`,
      });
    } catch (error) {
      Alert.alert('Error', error.message);
    }
  };

  // To avoid heavy library such as react-navigation, the screen with JSON tree
  // is displayed by using the showJSON***Details state for the
  // corresponding instance (Action or request). This screen can have five (5) differents views

  // 1. We want to display the action Redux in a JSON Tree
  if (showJSONActionDetails && props.item instanceof ReduxAction) {
    return (
      <JSONDetails
        title="Action details"
        onPressBack={() => setShowJSONActionDetails(false)}
        data={props.item?.action}
      />
    );
  }

  // 2. We want to display the a request RESPONSE BODY only in a JSON Tree
  if (showJSONResponseDetails && (props.item instanceof RNRequest || props.item instanceof NRequest)) {
    return (
      <JSONDetails
        title="Response details"
        onPressBack={() => setShowJSONResponseDetails(false)}
        data={props.item?.response}
      />
    );
  }

  // 3. We want to display the a request REQUEST BODY only in a JSON Tree
  if (showJSONRequestDetails && (props.item instanceof RNRequest || props.item instanceof NRequest)) {
    return (
      <JSONDetails
        title="Request details"
        onPressBack={() => setShowJSONRequestDetails(false)}
        data={props.item?.dataSent}
      />
    );
  }

  // 4. We want to display ALL INFORMATIONS in a redux action
  // At this time (for point 4 and 5), we store the result in _content variable
  // to be showed later.
  if (props.item instanceof ReduxAction) {
    // props.item.action should only contains 2 elements, a type and a payload (not necessary called payload).
    // In consequence, if it's not the type, that could be the payload
    const _infos = Object.entries(props.item.action).filter(value => value[0] !== 'type');
    const _reduxAction = {
      label: (_infos.length > 0 && _infos[0][0]) || undefined,
      payload: (_infos.length > 0 && _infos[0][1]) || undefined,
    };

    _action = () => _onShareReduxAction(props.item as ReduxAction);
    _content = (
      <ActionDetails
        item={props.item}
        reduxAction={_reduxAction}
        onPressViewMore={() => setShowJSONActionDetails(true)}
      />
    );
  }

  // 5. We want to display ALL INFORMATIONS in a request (React-Native or Native)
  if (props.item instanceof RNRequest || props.item instanceof NRequest) {
    _action = () => _onShareRequest(props.item as RNRequest | NRequest);
    _content = (
      <RequestDetails
        item={props.item}
        onPressViewMoreRequest={() => setShowJSONRequestDetails(true)}
        onPressViewMoreResponse={() => setShowJSONResponseDetails(true)}
        setSnackBarMessage={setSnackBarMessage}
        setSnackBarVisibility={setSnackBarVisibility}
      />
    );
  }

  // Appbar header is repeated here cause we use the absolute position in the style
  // Put this directly in the index.tsx cause that the Appbar will be added two times.
  return (
    <View style={styles.container}>
      <Appbar.Header style={[styles.header, { backgroundColor: theme.secondaryDarkColor }]}>
        <TouchableOpacity
          style={[styles.button, { borderLeftWidth: 0 }]}
          onPress={() => props.onPressBack(false)}
          testID="buttonBackToMainScreen"
        >
          <FeatherIcon name="arrow-left" color={theme.textColorOne} size={24} />
        </TouchableOpacity>
        <Appbar.Content color={theme.primaryColor} title="Netwatch" titleStyle={{ fontSize: 18 }} />
        <TouchableOpacity
          testID="buttonShare"
          style={[styles.button, { borderLeftWidth: 0 }]}
          onPress={() => _action()}
        >
          <FeatherIcon name="download" color={theme.textColorOne} size={24} />
        </TouchableOpacity>
      </Appbar.Header>

      <ScrollView style={{ backgroundColor: theme.secondaryColor }} contentContainerStyle={styles.scrollview}>
        {_content}
      </ScrollView>
      <Snackbar
        visible={snackBarVisibility}
        onDismiss={() => setSnackBarVisibility(false)}
        duration={3000}
        style={{ backgroundColor: theme.textColorOne }}
        theme={{
          colors: {
            surface: theme.textColorFour,
          },
        }}
      >
        {snackBarMessage}
      </Snackbar>
    </View>
  );
};

const styles = StyleSheet.create({
  header: {
    elevation: 0,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0,
  },

  container: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    top: 0,
    left: 0,
  },

  scrollview: {
    paddingTop: 26,
    paddingBottom: 20,
  },

  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
});
