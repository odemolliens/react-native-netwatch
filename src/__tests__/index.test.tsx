import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { IProps, Netwatch } from '../index';
import { Main } from '../Components/Main';
import { Modal } from 'react-native';
import { render, waitFor } from '@testing-library/react-native';

describe('Index test suite', () => {
  let component: ShallowWrapper;
  let props: IProps;
  let useEffect;
  let mockUseEffect;
  const setVisible = jest.fn();
  const onPressClose = jest.fn();
  const setShowDetails = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    mockUseEffect = () => {
      useEffect.mockImplementationOnce(f => f());
    };

    jest.mock('react', () => ({
      ...jest.requireActual('react'),
      useState: jest.fn(),
    }));
  });

  const globalDateConstructor = Date.now;

  beforeAll(() => {
    global.Date.now = jest.fn(() => new Date('2019-04-07T10:20:30Z').getTime());
  });

  afterAll(() => {
    global.Date.now = globalDateConstructor;
  });

  it('should render properly', () => {
    givenProps(true, true);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should call setVisible', () => {
    let useStateMock: any = (visible: any) => [false, setVisible];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    givenProps(false, true);
    mockUseEffect();
    givenComponent();
    mockUseEffect();
    expect(setVisible).toHaveBeenCalledTimes(1);
  });

  it('should contains the main screen with flatlist', async () => {
    // @ts-ignore
    const { getByTestId } = render(<Netwatch enabled visible reduxActions={[]} />);
    waitFor(() => expect(getByTestId('mainScreen')).toBeDefined());
    expect(getByTestId('itemsList')).toBeDefined();
  });

  it('should contains the details screen', async () => {
    // @ts-ignore
    const { getByTestId } = render(<Netwatch enabled reduxActions={[]} />);
    waitFor(() => expect(getByTestId('detailsScreen')).toBeDefined());
  });

  it('should disable netwatch', async () => {
    givenProps(false, false);
    givenComponent();
    expect(component).toEqual({});
  });

  it('should render properly visible and enabled', () => {
    givenProps(true, true);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should render properly visible, enabled and maxRequest', () => {
    givenProps(true, true, 20);
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should render properly Main component and close main component called setVisible', () => {
    let useStateMock: any = (visible: any) => [visible, setVisible];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    givenProps(true, true, 20);
    givenComponent();
    expect(component.find(Main)).toHaveLength(1);
    component.find(Main).invoke('onPressClose')(false);
    expect(setVisible).toHaveBeenCalledTimes(1);
  });

  it('should render properly Modal and called setVisible when back is pressed', () => {
    // Case when shake is actived
    let useStateMock: any = (visible: any) => [visible, setVisible];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    givenProps(true, true, 20);
    givenComponent();
    expect(component.find(Modal)).toHaveLength(1);
    component.find(Modal).invoke('onRequestClose')();
    expect(setVisible).toHaveBeenCalledTimes(1);
  });

  it('should render properly Modal and called props.onPressClose when back is pressed', () => {
    // Case when button is actived and shake desactivated
    givenProps(true, true, 20, true, true, 'dark', onPressClose);
    givenComponent();
    expect(component.find(Modal)).toHaveLength(1);
    component.find(Modal).invoke('onRequestClose')();
    expect(props.onPressClose).toHaveBeenCalledTimes(1);
  });

  it('should render properly Modal and called setShowDetails when back is pressed', () => {
    // Case when the Details page is visible
    let useStateMock: any = (showDetails: any) => [true, setShowDetails];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    givenProps(true, true, 20, true, true, 'dark', onPressClose);
    givenComponent();
    expect(component.find(Modal)).toHaveLength(1);
    component.find(Modal).invoke('onRequestClose')();
    expect(setShowDetails).toHaveBeenCalledTimes(1);
  });

  // GIVEN
  function givenComponent() {
    component = shallow(<Netwatch {...props} />);
  }

  function givenProps(
    visible: boolean = false,
    enabled: boolean = false,
    maxRequests?: number,
    disableShake: boolean = false,
    interceptIOS: boolean = true,
    theme: 'dark' | 'light' = 'dark',
    onPressClose = null,
  ) {
    props = {
      visible,
      enabled,
      maxRequests,
      disableShake,
      interceptIOS,
      theme,
      onPressClose,
    };
  }
});
