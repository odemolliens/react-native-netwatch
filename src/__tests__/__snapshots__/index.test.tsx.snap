// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index test suite should render properly 1`] = `
<ContextProvider
  value={
    Object {
      "base00": "#1F2937",
      "base03": "#06B6D4",
      "base07": "#D1D5DB",
      "base08": "#B91C1C",
      "base09": "#D97706",
      "base0B": "#4D7C0F",
      "base0D": "#F9FAFB",
      "black": "#000",
      "failedColor": "#B91C1C",
      "primaryColor": "#06B6D4",
      "reduxColor": "#764ABC",
      "secondaryColor": "#1F2937",
      "secondaryDarkColor": "#111827",
      "secondaryLightColor": "#374151",
      "successColor": "#4D7C0F",
      "textColorFour": "#6B7280",
      "textColorOne": "#F9FAFB",
      "textColorThree": "#9CA3AF",
      "textColorTwo": "#D1D5DB",
      "warningColor": "#D97706",
      "white": "#FFF",
    }
  }
>
  <Component
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    visible={false}
  >
    <Component
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Component
        style={
          Object {
            "height": "100%",
          }
        }
      >
        <Main
          clearAll={[Function]}
          connections={Array []}
          maxRequests={100}
          nRequests={Array []}
          onPress={[Function]}
          onPressClose={[Function]}
          onPressDetail={[Function]}
          reduxActions={Array []}
          rnRequests={Array []}
          testId="mainScreen"
        />
      </Component>
      <Component
        style={
          Object {
            "height": 0,
          }
        }
      >
        <Details
          item={
            ReduxAction {
              "_id": -1,
              "action": Object {
                "payload": "",
                "type": "__ERROR:UNDEFINED__",
              },
              "startTime": 1554632430000,
              "stringifiedAction": "",
              "type": "REDUX",
            }
          }
          onPressBack={[Function]}
          testId="detailScreen"
        />
      </Component>
    </Component>
  </Component>
</ContextProvider>
`;

exports[`Index test suite should render properly visible and enabled 1`] = `
<ContextProvider
  value={
    Object {
      "base00": "#1F2937",
      "base03": "#06B6D4",
      "base07": "#D1D5DB",
      "base08": "#B91C1C",
      "base09": "#D97706",
      "base0B": "#4D7C0F",
      "base0D": "#F9FAFB",
      "black": "#000",
      "failedColor": "#B91C1C",
      "primaryColor": "#06B6D4",
      "reduxColor": "#764ABC",
      "secondaryColor": "#1F2937",
      "secondaryDarkColor": "#111827",
      "secondaryLightColor": "#374151",
      "successColor": "#4D7C0F",
      "textColorFour": "#6B7280",
      "textColorOne": "#F9FAFB",
      "textColorThree": "#9CA3AF",
      "textColorTwo": "#D1D5DB",
      "warningColor": "#D97706",
      "white": "#FFF",
    }
  }
>
  <Component
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    visible={false}
  >
    <Component
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Component
        style={
          Object {
            "height": "100%",
          }
        }
      >
        <Main
          clearAll={[Function]}
          connections={false}
          maxRequests={100}
          nRequests={false}
          onPress={[MockFunction]}
          onPressClose={[Function]}
          onPressDetail={[MockFunction]}
          reduxActions={false}
          rnRequests={false}
          testId="mainScreen"
        />
      </Component>
      <Component
        style={
          Object {
            "height": 0,
          }
        }
      >
        <Details
          item={false}
          onPressBack={[MockFunction]}
          testId="detailScreen"
        />
      </Component>
    </Component>
  </Component>
</ContextProvider>
`;

exports[`Index test suite should render properly visible, enabled and maxRequest 1`] = `
<ContextProvider
  value={
    Object {
      "base00": "#1F2937",
      "base03": "#06B6D4",
      "base07": "#D1D5DB",
      "base08": "#B91C1C",
      "base09": "#D97706",
      "base0B": "#4D7C0F",
      "base0D": "#F9FAFB",
      "black": "#000",
      "failedColor": "#B91C1C",
      "primaryColor": "#06B6D4",
      "reduxColor": "#764ABC",
      "secondaryColor": "#1F2937",
      "secondaryDarkColor": "#111827",
      "secondaryLightColor": "#374151",
      "successColor": "#4D7C0F",
      "textColorFour": "#6B7280",
      "textColorOne": "#F9FAFB",
      "textColorThree": "#9CA3AF",
      "textColorTwo": "#D1D5DB",
      "warningColor": "#D97706",
      "white": "#FFF",
    }
  }
>
  <Component
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    visible={false}
  >
    <Component
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Component
        style={
          Object {
            "height": "100%",
          }
        }
      >
        <Main
          clearAll={[Function]}
          connections={false}
          maxRequests={20}
          nRequests={false}
          onPress={[MockFunction]}
          onPressClose={[Function]}
          onPressDetail={[MockFunction]}
          reduxActions={false}
          rnRequests={false}
          testId="mainScreen"
        />
      </Component>
      <Component
        style={
          Object {
            "height": 0,
          }
        }
      >
        <Details
          item={false}
          onPressBack={[MockFunction]}
          testId="detailScreen"
        />
      </Component>
    </Component>
  </Component>
</ContextProvider>
`;
